internal/configmodels/configmodels.go

```diff
@@ -0,0 +1,157 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package configmodels defines the data models for entities. This file defines the
+// models for V2 configuration format. The defined entities are:
+// ConfigV2 (the top-level structure), Receivers, Exporters, Processors, Pipelines.
+package configmodels
+
+/*
+Receivers, Exporters and Processors typically have common configuration settings, however
+sometimes specific implementations (e.g. StackDriver) will have extra configuration settings.
+This requires the configuration data for these entities to be polymorphic.
+
+To satisfy these requirements we declare interfaces ReceiverCfg, ExporterCfg, ProcessorCfg,
+which define the behavior. We also provide helper structs ReceiverCommon, ExporterCommon,
+ProcessorCommon, which define the common settings and un-marshaling from config files.
+
+Specific Receivers/Exporters/Processors are expected to at the minimum implement the
+corresponding interface and if they have additional settings they must also extend
+the corresponding common settings struct (the easiest approach is to embed the common struct).
+*/
+
+// ConfigV2 defines the configuration V2 for the various elements of collector or agent.
+type ConfigV2 struct {
+	Receivers  []ReceiverCfg  `mapstructure:"-"`
+	Exporters  []ExporterCfg  `mapstructure:"-"`
+	Processors []ProcessorCfg `mapstructure:"-"`
+	Pipelines  Pipelines      `mapstructure:"pipelines"`
+}
+
+// ReceiversKeyName is the configuration key name for receivers section.
+const ReceiversKeyName = "receivers"
+
+// ExportersKeyName is the configuration key name for exporters section.
+const ExportersKeyName = "exporters"
+
+// ProcessorsKeyName is the configuration key name for processors section.
+const ProcessorsKeyName = "processors"
+
+// Pipelines defines the pipelines section of configuration.
+type Pipelines struct {
+	Traces  []*Pipeline `mapstructure:"traces"`
+	Metrics []*Pipeline `mapstructure:"metrics"`
+}
+
+// Pipeline defines a single pipeline.
+type Pipeline struct {
+	Name       string              `mapstructure:"name"`
+	Receivers  []PipelineReceiver  `mapstructure:"receivers"`
+	Processors []PipelineProcessor `mapstructure:"ordered-processors"`
+	Exporters  []PipelineExporter  `mapstructure:"exporters"`
+}
+
+// PipelineReceiver defines a receiver of a pipeline.
+type PipelineReceiver struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineProcessor defines a processor of a pipeline.
+type PipelineProcessor struct {
+	Name string `mapstructure:"name"`
+}
+
+// PipelineExporter defines an exporter of a pipeline.
+type PipelineExporter struct {
+	Name string `mapstructure:"name"`
+}
+
+// NamedEntity defines a configuration entity that has a name.
+type NamedEntity interface {
```

Comment:
Re-opening an issue that we did not conclude in the previous PR:
https://github.com/census-instrumentation/opencensus-service/pull/543#discussion_r283521939

I think composition should be enough here. Can you explain where you need more?
I like this article about inheritance in Go https://medium.com/@simplyianm/why-gos-structs-are-superior-to-class-based-inheritance-b661ba897c67

Reply:
Perhaps I misundersand your question but let me try to answer.

First of all, you are right that I do not _need_ NamedEntity to be an interface. I can eliminate it and simply have `GetName()` function in all interfaces which must have a name.

However, I believe using an embedded interface is idiomatic Go and is widely used in Go standard libraries as well. See section about embedding interfaces here: https://golang.org/ref/spec#Interface_types

I would like to understand why you think this is a bad idea. Happy to eliminate it if you provide the reasoning.

Or are you asking about why I need polymorphism in the configuration at all?

Reply:
One more thing: I do not have any inheritance here (strictly speaking there is no struct inheritance in Go, you can only compose structs, you can implement interfaces and you can embed interfaces and structs).

In this PR I achieve polymorphism in behavior via interfaces which is as far as I know the idiomatic way to do polymorphism in Go.

I also read the article that you linked to. I am using exactly what is described in the article (i.e. interface and struct embedding - https://golang.org/doc/effective_go.html#embedding).

Reply:
Ok I see the mess that Go has here :). I was mostly curious if we can enforce that every config has a:
```
Name    string `mapstructure:"name"`
```

But probably it is harder to achieve.

Reply:
So you are looking for a way to make sure that every config's name field is defined in a way that is consistent. So that for example we do not have a Receiver config which for example uses "namestring" as field name :-)
Is that what you are aiming for? I agree it is a good goal, let me think.

Reply:
Yes.

Reply:
Hmm, I cannot think of a nice way to achieve this. The only thing I can suggest is what I already had: to have a `ReceiverCommon` struct which contains the _recommended_ marshaling layout. Of course it does not enforce anything but at least it provides a guideline to implementors:

```go
type ReceiverCommon struct {
	Type     string `mapstructure:"type"`
	Enabled  bool   `mapstructure:"enabled"`
	Name     string `mapstructure:"name"`
	Endpoint string `mapstructure:"endpoint"`
}
```